### DISCUSSION TITLE
Vite Plugin: Support `server` property from remix.config.js to enable @remix-run/architect

### DISCUSSION THREAD

**Discussion by whanso** [remix-run/remix#8836] [Q&A]
Recently I was able to get a deployment to AWS enabled using `@remix-run/architect` to configure a handler and CDK to stand up AWS architecture (not using Architect for deployment).

I'd love to incorporate Vite into the project; currently, the `@remix-run/architect` Remix template has not been renovated to use Vite, so I'm trying to figure it out.  See the template: https://github.com/remix-run/remix/tree/main/templates/arc

The codebase for `@remix-run/architect` requires a server.ts file to implement the `createRequestHandler`.  To point Remix to the file, historically we do that as:

```js
/** @type {import('@remix-run/dev').AppConfig} */
export default {
  server: "server.ts",
  ...
};
```

According to the docs, the config options for `remix.config.ts` should carry over to the Vite plugin, but alas, I do not see the `server` property to bridge architect over correctly.  

<img width="474" alt="image" src="https://github.com/remix-run/remix/assets/26184716/fcdcae91-b755-4dcb-92a9-243aca8798c9">

Let me know if I'm heading in the right direction, but it seems like some additional enabling will be necessary to get `@remix-run/architect` to play nicely with the new Vite plugin.

**Comment by MichaelDeBoey** [comment]
If you follow the ["Migrating a custom server" section in the docs](https://remix.run/future/vite#migrating-a-custom-server), you should be able to get it working again

**Reply by sergiodxa** [reply]
It would be great if Vite could also compile the custom server when using TS instead of using `node --loader tsm ./server.ts` or `tsx ./server.ts`

**Reply by whanso** [reply]
Hi @MichaelDeBoey , thanks, I will give that a look.

**Comment by MattyBalaam** [comment]
@MichaelDeBoey I could do with a little more help, the section you linked just seems to be about development, but we only use a custom server when we build.

In our previous config we used our custom server for AWS when building for prod:

```
module.exports = {
 server:
    process.env.NODE_ENV === 'development'
      ? undefined
      : 'path-to-custom-server.ts'
      …
``` 

But I’m not sure how we can change the new `vite:build` to reference our file which uses `createRequestHandler` from this package: ` https://www.npmjs.com/package/remix-aws`

**Reply by whanso** [reply]
@MattyBalaam @MichaelDeBoey I played with it a little more myself, and admit that the docs for refactoring the express example is not a true carryover.  I made a guess, but I was not successful in starting the server locally using the `createRequestHandler` from the architect package.

**Reply by MattyBalaam** [reply]
Digging around it seems that you maybe need to add a magic file in the root of the package called server.js (or server.ts if you add a script called `start` in `package.json` ?

I’m not sure though it does not seem very well documented

**Comment by MattyBalaam** [comment]
@whanso I have a feeling that there needs to be a vitePlugin made for this, something like this: https://github.com/remix-run/remix/blob/main/packages/remix-dev/vite/cloudflare-proxy-plugin.ts which is used in: https://github.com/remix-run/remix/blob/main/templates/vite-cloudflare/vite.config.ts

**Reply by whanso** [reply]
I think it's just a matter of time until the official adapters are updated to support the Vite plugin.  Maybe something supplemental will be needed, though, but it will call for modifying the `server.ts` file to incorporate something from Vite.

**Reply by MattyBalaam** [reply]
OK, in our case we are not using architect, but maybe this can help how I got it working:

In our particular use case we were running our app using `NodejsFunction`from `aws-cdk-lib/aws-lambda-nodejs` and we were using the `server` entrypoint option in our `remix.config` file to point to our `server.ts` file which exports a `handler` function created by remix-aws like this:

```
import * as build from '@remix-run/dev/server-build';

export const handler = createRequestHandler({
  build,
  mode: process.env.NODE_ENV,
  awsProxy: AWSProxy.APIGatewayV1,
});
```

We then pointed our `NodejsFunction` to the output of the remix build.

However, we have now flipped this around. Insteead we let remix build, and then we import the build path into our `server.ts` file:

```
export const handler = createRequestHandler({
  // @ts-expect-error this may not exist at all times so will throw an error - awaiting help from remix team
  build: await import('./build/server/index.js'),
  mode: process.env.NODE_ENV,
  awsProxy: AWSProxy.APIGatewayV1,
});
```

And then finallay in our `NodejsFunction` we point directly at the `server.ts` file (instead of the build output)

**Reply by owenmorgan** [reply]
> OK, in our case we are not using architect, but maybe this can help how I got it working:
> 
> In our particular use case we were running our app using `NodejsFunction`from `aws-cdk-lib/aws-lambda-nodejs` and we were using the `server` entrypoint option in our `remix.config` file to point to our `server.ts` file which exports a `handler` function created by remix-aws like this:
> 
> ```
> import * as build from '@remix-run/dev/server-build';
> 
> export const handler = createRequestHandler({
>   build,
>   mode: process.env.NODE_ENV,
>   awsProxy: AWSProxy.APIGatewayV1,
> });
> ```
> 
> We then pointed our `NodejsFunction` to the output of the remix build.
> 
> However, we have now flipped this around. Insteead we let remix build, and then we import the build path into our `server.ts` file:
> 
> ```
> export const handler = createRequestHandler({
>   // @ts-expect-error this may not exist at all times so will throw an error - awaiting help from remix team
>   build: await import('./build/server/index.js'),
>   mode: process.env.NODE_ENV,
>   awsProxy: AWSProxy.APIGatewayV1,
> });
> ```
> 
> And then finallay in our `NodejsFunction` we point directly at the `server.ts` file (instead of the build output)

can you share some more around this? where does AWSProxy come from?

**Reply by MattyBalaam** [reply]
> > OK, in our case we are not using architect, but maybe this can help how I got it working:
> > In our particular use case we were running our app using `NodejsFunction`from `aws-cdk-lib/aws-lambda-nodejs` and we were using the `server` entrypoint option in our `remix.config` file to point to our `server.ts` file which exports a `handler` function created by remix-aws like this:
> > ```
> > import * as build from '@remix-run/dev/server-build';
> > 
> > export const handler = createRequestHandler({
> >   build,
> >   mode: process.env.NODE_ENV,
> >   awsProxy: AWSProxy.APIGatewayV1,
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We then pointed our `NodejsFunction` to the output of the remix build.
> > However, we have now flipped this around. Insteead we let remix build, and then we import the build path into our `server.ts` file:
> > ```
> > export const handler = createRequestHandler({
> >   // @ts-expect-error this may not exist at all times so will throw an error - awaiting help from remix team
> >   build: await import('./build/server/index.js'),
> >   mode: process.env.NODE_ENV,
> >   awsProxy: AWSProxy.APIGatewayV1,
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And then finallay in our `NodejsFunction` we point directly at the `server.ts` file (instead of the build output)
> 
> can you share some more around this? where does AWSProxy come from?

Sorry, this was from a helper package remix-aws: https://www.npmjs.com/package/remix-aws. See: https://github.com/wingleung/remix-aws/blob/main/src/server.ts it is actually a string `APIGatewayV1`

**Reply by gotgenes** [reply]
Following the discussions above, and digging deep into Vite and Rollup documentation, here is my solution:

In `vite.config.ts`:

```typescript
import { vitePlugin as remix } from "@remix-run/dev";
import { installGlobals } from "@remix-run/node";
import { defineConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";

installGlobals();

export default defineConfig({
  plugins: [
    remix(),
    tsconfigPaths(),
    {
      name: "remix-apigatewayv2-adapter",
      apply(_config, env): boolean {
        return env.command === "build" && env?.isSsrBuild === true;
      },
      config: async (_config, _env) => {
        return {
          build: {
            ssr: "handler.ts",
          },
        };
      },
    },
  ],
});
```

Here is `handler.ts`, which is at the root level of the project (outside of `app/`, and alongside `vite.config.ts`):

```typescript
import { AWSProxy, createRequestHandler } from "remix-aws";
import * as build from "virtual:remix/server-build";

export const handler = createRequestHandler({
  build: build,
  mode: process.env.NODE_ENV,
  awsProxy: AWSProxy.APIGatewayV2,
});
```

Note that `tsc` will complaine that it doesn't know what the types are from `"virtual:remix/server-build"`, because Remix does not expose types for this [virtual module defined in its Vite plugin](https://github.com/remix-run/remix/blob/393d70b59d7d24cc6d2655bdd307c5daba4a4b82/packages/remix-dev/vite/plugin.ts#L759-L821).

`createRequestHandler` is from [the remix-aws third-party adapter library](https://github.com/wingleung/remix-aws).

**Reply by RoelanddeBetue** [reply]
Thanks @gotgenes I got it working with the vite config! I'm using a createRequestHandler for apiGatewayV1 that I wrote myself based on the third-party adapter.

**Reply by andrewdavidcostello** [reply]
@gotgenes nice sleuthing, thanks! (saved me a headache) I hope somebody from the remix team can address this sooner rather than later to at least supply the types and state if this should be an accepted work around for everyone trying to run remix on AWS and not wanting to use the amplify method.

Does anybody know if there is an issue to track this as would be good to propose the above as a solution and ask the team for a more robust fix?

**Reply by davisg123** [reply]
@gotgenes What does your `dev` command look like in this case? I'm a bit confused about how the architect sandbox is run with this solution

**Reply by charlee-dev** [reply]
i would think that this solution does not use the architect at all, but i also may be wrong.. ;)

**Reply by charlee-dev** [reply]
i've got a small [sample](https://github.com/charlee-dev/cdk-remix-vite-sample) which is still using the architect.

**Reply by gotgenes** [reply]
I just published [a repository to demonstrate deploying to AWS Lambda](https://github.com/gotgenes/remix-aws-lambda-app), based on [the solution I described above](https://github.com/remix-run/remix/discussions/8836#discussioncomment-9795687). Hopefully this helps others.

**Comment by vinarius** [comment]
I've sunk a good 12 hours into trying to get a standalone build to export into a lambda function, no success.

I think I would have loved building my next application with remix, but I am choosing to abandon it and go back to next. Much of the docs are out of date since the move to vite. Hopefully this will receive more attention in the future.

**Reply by ch-stuart** [reply]
I've had pretty good luck with [SST](https://sst.dev).

**Comment by vinarius** [comment]
Wow thanks! I'll have to take a closer look at this when I get some time.

**Comment by quantuminformation** [comment]
So glad to see the discussion going on here I'm ready to take the bite and migrate a ark project to vite
